app/src/main/java/com/internship/move/presentation/map/MapFragment.kt

```diff
@@ -1,35 +1,114 @@
+import androidx.activity.result.contract.ActivityResultContracts
+import androidx.core.content.ContextCompat
 import androidx.fragment.app.Fragment
 import androidx.navigation.fragment.findNavController
+import com.google.android.gms.location.FusedLocationProviderClient
+import com.google.android.gms.location.LocationServices
+import com.google.android.gms.maps.CameraUpdateFactory
+import com.google.android.gms.maps.GoogleMap
+import com.google.android.gms.maps.OnMapReadyCallback
+import com.google.android.gms.maps.SupportMapFragment
+import com.google.android.gms.maps.model.LatLng
+import com.google.android.gms.maps.model.MapStyleOptions
+import com.google.android.gms.maps.model.MarkerOptions
+import com.google.android.gms.tasks.Task
 import com.internship.move.R
 import com.internship.move.databinding.FragmentMapBinding
 import com.internship.move.presentation.map.viewmodel.MapViewModel
 import com.zhuinden.fragmentviewbindingdelegatekt.viewBinding
 import org.koin.androidx.viewmodel.ext.android.viewModel
 
-class MapFragment : Fragment(R.layout.fragment_map) {
+class MapFragment : Fragment(R.layout.fragment_map), OnMapReadyCallback {
 
     private val binding by viewBinding(FragmentMapBinding::bind)
     private val viewModel by viewModel<MapViewModel>()
+    private lateinit var map: GoogleMap
+    private lateinit var fusedLocationProviderClient: FusedLocationProviderClient
 
     override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
         super.onViewCreated(view, savedInstanceState)
 
-        binding.mapFragmentLogoutBtn.setOnClickListener {
-            viewModel.logOut()
-            findNavController().navigate(MapFragmentDirections.actionMapFragmentToSplashGraph())
+        val mapFragment = childFragmentManager.findFragmentById(R.id.mapFragment) as SupportMapFragment
+        fusedLocationProviderClient = LocationServices.getFusedLocationProviderClient(requireActivity())
+
+        permissionInit(mapFragment)
+        initListeners()
+    }
+
+    private fun initListeners() {
+        binding.menuIV.setOnClickListener {
+            findNavController().navigate(MapFragmentDirections.actionMapFragmentToMenuFragment())
+        }
+        binding.locationIV.setOnClickListener {
+            fetchLocation()
+        }
+    }
+
+    private fun permissionInit(mapFragment: SupportMapFragment) {
+        val requestPermissionLauncher = ContextCompat.checkSelfPermission(requireContext(), Manifest.permission.ACCESS_FINE_LOCATION)
+
+        if (requestPermissionLauncher != PackageManager.PERMISSION_GRANTED) {
+            permissionsResultCallback.launch(Manifest.permission.ACCESS_FINE_LOCATION)
+        } else {
+            binding.locationIV.setImageResource(R.drawable.ic_focus_location)
+            binding.headingTV.setText(R.string.map_header_location_allowed)
+            mapFragment.getMapAsync(this)
+        }
+    }
+
+    private val permissionsResultCallback = registerForActivityResult(
```

Comment:
Keep fields together at the beginning of the class and implement a `initPermissionsResultCallback()` method to be called there